### PR TITLE
Fix config override of other settings levels

### PR DESCRIPTION
**Pull request by langleyd**
Fixes https://github.com/element-hq/element-web/issues/26877

### CODE REVIEW

**Review by t3chguy — COMMENTED** [review]
It feels strange to have 2 mechanisms which can make config take priority. `supportedLevels` being the other one as the levels are ordered IIRC

**Comment by langleyd** [comment]
Yea there is a `supportedLevelsAreOrdered` which instructs that `supportedLevels ` should be respected in order.
 
This PR was my attempt to have `configDisablesSetting` make a bit more sense. As i'm not sure disabling a setting for a user where there is a value in config makes sense without also enforcing that the config value is actually used(overrides the local one).


@t3chguy For all the settings with that have`configDisablesSetting` we could instead use `supportedLevelsAreOrdered` and use a level order with config coming first. In this case would we expect to still need the explicit `configDisablesSetting` flag to disabled the setting in the UI, so it always needs to be used alongside `supportedLevelsAreOrdered` ? Or would you expect we add a check to see if the current setting is being overridden by config based on `supportedLevelsAreOrdered` and that causes the setting UI to be disabled?

**Comment by t3chguy** [comment]
> Or would you expect we add a check to see if the current setting is being overridden by config based on supportedLevelsAreOrdered and that causes the setting UI to be disabled?

This feels like a nice way of doing it IMO, condensing the options we have and doing more stuff implicitly

File: src/settings/SettingsStore.ts [hunk 1]
```diff
@@ -505,32 +505,32 @@ export default class SettingsStore {
...
-        if (SETTINGS[settingName].controller?.settingDisabled) {
+        if (setting.controller?.settingDisabled) {
             return false;
         }
 
-        // For some config settings (mostly: non-beta features), a value in config.json overrides the local setting
-        // (ie: we force them as enabled or disabled).
-        if (SETTINGS[settingName]?.configDisablesSetting) {
+        // // For some config settings (mostly: non-beta features), a value in config.json overrides the local setting
+        // // (ie: we force them as enabled or disabled). In this case we should not let the user change the setting.
```
Comment (t3chguy) [inline src/settings/SettingsStore.ts:532]: double comment?

File: src/components/views/elements/SettingsFlag.tsx [hunk 2]
```diff
@@ -62,6 +62,17 @@ export default class SettingsFlag extends React.Component<IProps, IState> {
     }
 
     private getSettingValue(): boolean {
+        // If a level defined in props is overridden by level at a high presenence, it gets disabled
```
Comment (t3chguy) [inline src/components/views/elements/SettingsFlag.tsx:65]: typo

File: src/components/views/elements/SettingsFlag.tsx [hunk 3]
```diff
@@ -96,7 +107,6 @@ export default class SettingsFlag extends React.Component<IProps, IState> {
 
     public render(): React.ReactNode {
         const disabled = !SettingsStore.canSetValue(this.props.name, this.props.roomId ?? null, this.props.level);
-
```
Comment (t3chguy) [inline src/components/views/elements/SettingsFlag.tsx:99]: unrelated

**Review by t3chguy — APPROVED** [review]

**Comment by langleyd** [comment]
@t3chguy I'm not sure disabling settings that have one set at a high level, necessarily makes sense in all cases. [blacklistUnverifiedDevices](https://github.com/matrix-org/matrix-react-sdk/blob/develop/src/settings/Settings.tsx#L913) is exposed in in room and device settings. And while room should take precedence in this case, both should be enabled.

By chance, this code didn't break things as when checking `canSetValue` at the device level you don't have the roomId, so doesn't see a value set at a higher level.

I wonder should we have a "higherPrecenceValuesShouldDisableSetting" property that is true by default but can be set to false?

**Comment by t3chguy** [comment]
It should probably be special-cased to `config` only as that's the only read-only setting level

**Comment by langleyd** [comment]
This seem ok now @t3chguy ?

File: src/settings/SettingsStore.ts [hunk 4]
```diff
@@ -505,41 +505,70 @@ export default class SettingsStore {
...
+        level: SettingLevel,
+    ): boolean {
+        const setting = SETTINGS[settingName];
+        const levelOrders = getLevelOrder(setting);
+        const configIndex = levelOrders.indexOf(SettingLevel.CONFIG);
+        const levelIndex = levelOrders.indexOf(level);
+        console.log("settingIsOveriddenAtConfigLevel");
+        console.log(settingName);
+        console.log(configIndex);
+        console.log(levelIndex);
```
Comment (t3chguy) [inline src/settings/SettingsStore.ts:564]: These should probably not land

File: src/components/views/elements/SettingsFlag.tsx [hunk 5]
```diff
@@ -62,6 +62,17 @@ export default class SettingsFlag extends React.Component<IProps, IState> {
     }
 
     private getSettingValue(): boolean {
+        // If a level defined in props is overridden by a level at a high presedence, it gets disabled
+        // and we should show the overridding value.
+        if (
+            !!SettingsStore.settingIsOveriddenAtConfigLevel(
```
Comment (t3chguy) [inline src/components/views/elements/SettingsFlag.tsx:68]: don't need the `!!` here given its a boolean

Comment (t3chguy) [reply src/components/views/elements/SettingsFlag.tsx:68]: This is identified by Sonar: https://sonarcloud.io/project/issues?id=matrix-react-sdk&pullRequest=12593&resolved=false&sinceLeakPeriod=true
